### PR TITLE
feat: Add entity output hooking

### PR DESCRIPTION
**Pull request by Poggicek**
This pull requests implements the ability to hook entity outputs for a given classname or entity.

# Examples

## HookEntityOutput

```c#
public override void Load(bool hotReload)
{
    HookEntityOutput("weapon_knife", "OnPlayerPickup", (string name, CEntityInstance activator, CEntityInstance caller, float delay) =>
    {
        Logger.LogInformation("weapon_knife called OnPlayerPickup ({name}, {activator}, {caller}, {delay})", name, activator.DesignerName, caller.DesignerName, delay);
    });
}
```

## EntityOutputHook Attribute

```c#
[EntityOutputHook("weapon_knife", "OnPlayerPickup")]
public void OnKnifePickup(string name, CEntityInstance activator, CEntityInstance caller, float delay)
{
    Logger.LogInformation("[EntityOutputHook Attribute] weapon_knife called OnPlayerPickup ({name}, {activator}, {caller}, {delay})", name, activator.DesignerName, caller.DesignerName, delay);
}
```

## Hook single entity

```c#
CBaseDoor? door = Utilities.FindAllEntitiesByDesignerName<CBaseDoor>("prop_door_rotating").FirstOrDefault();

if (door)
    HookSingleEntityOutput(door, "OnOpen", OnOpen);

public void OnOpen(string name, CEntityInstance activator, CEntityInstance caller, float delay)
{
    Console.WriteLine($"{activator.DesignerName} called {name}");
    
    // fire only once
    UnhookSingleEntityOutput(caller, "OnOpen", OnOpen);
}
```

### CODE REVIEW

**Comment by roflmuffin** [comment]
Do we also want to be able to hook based on a reference to the existing schema outputio? So we can also support entity.OnDoorOpen as a passable reference for instance?

**Comment by roflmuffin** [comment]
Well imo all the struct information from cpp added in this PR should be added to the CEntityIOOutput anyway because a user may want to know the name/description of an output on the managed side

**Comment by KillStr3aK** [comment]
> How do we feel about making the entity output callbacks return a hookresult so they can cancel the underlying output call? I haven't tested what happens when you cancel a given output call though.

sounds good to me, might worth give it a try

File: managed/CounterStrikeSharp.API/Core/BasePlugin.cs [hunk 1]
```diff
@@ -430,6 +445,81 @@ public void RegisterConsoleCommandAttributeHandlers(object instance)
...
+            EntityOutputHooks.Remove(handler);
+        }
+
+        public void HookSingleEntityOutput(CEntityInstance entityInstance, string outputName, EntityIO.EntityOutputHandler handler)
+        {
+            // since we wrap around the plugin handler we need to do this to ensure that the plugin callback is only called
+            // if the entity instance is the same.
+            EntityIO.EntityOutputHandler internalHandler = (string name, CEntityInstance activator, CEntityInstance caller, float delay) =>
+            {
+                // TODO: this needs to be changed when equality check between 'CEntityInstance' instances is fixed
```
[the remark this answers is not in the document]
Comment (roflmuffin) [reply managed/CounterStrikeSharp.API/Core/BasePlugin.cs:489]: Yeah I tested it in the entity spawn handler by casting the entity as another type (smoke grenade), then doing all three kinds of equality checks, like you sent me and they all resolve to true 🙏

**Comment by Poggicek** [comment]
> Well imo all the struct information from cpp added in this PR should be added to the CEntityIOOutput anyway because a user may want to know the name/description of an output on the managed side. I can merge that in separately though

Should PR it into my fork then so that we can replace the name argument in the callback with the entityio object without breaking compatibility in the future

**Comment by Poggicek** [comment]
> Do we also want to be able to hook based on a reference to the existing schema outputio? So we can also support door.OnOpen as a passable reference for instance?

Not quite sure about this though, it would be easy to to do managed, but the question rather is, should we? I believe that we either keep the current HookSingleEntityOutput/UnhookSingleEntityOutput or we add .Hook, .Unhook into CEntityIOOutput but not both. There's very little reason to keep both and having multiple useless way to do one thing will only lead to confusion for other developers.

**Comment by Poggicek** [comment]
CEntityIOOutput.Hook would be better in the sense that you would have intellisense on every possible output for a given entity, however that does require you to cast the entity properly if it's just a CEntityInstance.

**Comment by roflmuffin** [comment]
> > Do we also want to be able to hook based on a reference to the existing schema outputio? So we can also support door.OnOpen as a passable reference for instance?
> 
> Not quite sure about this though, it would be easy to to do in managed, but the question rather is, should we? I believe that we either keep the current HookSingleEntityOutput/UnhookSingleEntityOutput or we add .Hook, .Unhook into CEntityIOOutput but not both. There's very little reason to keep both and having multiple useless way to do one thing will only lead to confusion for other developers.

I just mean an overload for `HookSingleEntityOutput` that takes an instance of `CEntityIOOutput` rather than a string, and then uses that CEntityIOOutput to get the name and pass it in as a string, just means that I am still using strings but at least if I pass it an instance of an output it still works.

**Comment by Poggicek** [comment]
Ah right I see, sure

**Comment by roflmuffin** [comment]
Sorry I oopsied this branch, could you re-open

**Pull request merged**